CHANGE: Add property drawer for InputActionsAsset to use Project-Wide Actions (ISX-1321)

jfreire-unity opened the pull request:
### Description

Implements https://jira.unity3d.com/browse/ISX-1321 for InputActionAsset. Follow-up PRs will update InputActionReference with project-wide actions.

### Changes made


![Kapture 2023-09-15 at 11 28 32](https://github.com/Unity-Technologies/InputSystem/assets/123556071/6c13b1f7-0313-4d39-b233-b2a4b32f4e1d)


This PR introduces a `CustomPropertyDrawer` for InputActionAsset. Whenever an InputActionAsset field is shown in the editor, it will show a dropdown menu to allow the selection for an actions asset or project-wide actions.
A "Open" button is also added when the user selects project wide actions.

 `InputSystemUIInputModule` is also updated so that the custom inspector editor shows the appropriate actions when using project-wide actions.


### Notes
- This is only available for 2022.3+, as it is related to project-wide actions.
- Fixes some bugs on the 'Input System UI Input Module' when working null asset selected, or missing actions in a valid asset.
- TrackedDevicePosition and TrackedDeviceOrientation added back into the project-wide actions in the UI action map so the default 'Input System UI Input Module' mappings function.

### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

jfreire-unity committed:
Add InputActionAssetProperty wrapper

This is necessary so that the Editor can hold the InputActionAsset settings to know if its project wide actions or not.

jfreire-unity committed:
WIP Add custom property drawer for InputActionAssetProperty

jfreire-unity committed:
Update PlayerInput component to use InputActionAssetProperty

jfreire-unity committed:
Cache properties and factor out option selection

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/InputActionAssetPropertyDrawer.cs:
```diff
@@ -0,0 +1,58 @@
+namespace UnityEngine.InputSystem.Editor
+{
+    // [CustomPropertyDrawer(typeof(InputActionAsset))]
+    [CustomPropertyDrawer(typeof(InputActionAssetProperty))]
```
It would be highly desirable to have this as [CustomPropertyDrawer(typeof(InputActionAsset))] instead. Then any third-party or other package code would just work, while this approach may enable assigning project-wide or asset it has the drawback of requiring anyone interested in referencing project wide to update/adapt their code. 

I did a quick test with CustomPropertyDrawer(typeof(InputActionAsset))] and that enables a drawer but leaves you with the problem of figuring out what is really selected to be able to set the value of the first high-level drop-down, which leads me to my next question:
How do we currently reference the project-wide asset? Is it available as a static instance somewhere? If it is, and I interpret InputActionAsset as basically being a reference (project wide is serialised independently by the settings?), why can't we leave PlayerInput and PlayerINputEditor unmodified and instead provide a new property on the InputActionAsset (evaluated by the drawer) which is bool IsProjectWide().
That getter could in turn base the result, ideally on reference comparison with the project wide instance, or on detecting a tag from json parsing (InputActionAsset.FromJson) not serialised for any other asset. What I am after is that any solution not introducing a new type here would be desirable in my perspective.

jfreire-unity replied:
My initial idea was to do a custom drawer for `InputActionAsset` but after not being able to make it work properly I decided with a different strategy, and found that this pattern would do what we wanted.
That idea sounds interesting and I'll try it out, thanks. 👍

lyndon-unity replied:
Good proposal and seems reasonable if this works.

jfreire-unity replied:
Addressed in newer commits. Thanks @ekcoh for the suggestion

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/InputActionAssetProperty.cs:
```diff
@@ -0,0 +1,12 @@
+namespace UnityEngine.InputSystem.Editor
+{
+    [Serializable]
+    public struct InputActionAssetProperty
```
Summary of problems I could see with this new type:
- This type is another type than InputActionAsset which implies we do not benefit from automatic property drawers on any existing code where InputActionAsset is a property, e.g. components in other packages or projects. Still I understand that you need a way to tell whether a certaain InputActionAsset is the global one or not, hence the bool.
- Given this solution what would happen if multiple InputActionAssetProperties exist which have the bool set to true but pointing to different assets? This type would allowing setting the bool from user code which may lead to  unexpected effects.

jfreire-unity replied:
Good points, thanks.
From my view, point 2 is the more critical from an API perspective.
We could probably leave with point 1 as we would only update our known components, leaving users to do what they wanted. Not desirable, but reasonable enough if there.
Based on your comments, I'll try again implementing this to `InputActionAsset` instead of creating a new type.

ekcoh reviewed (commented):
Only commenting for now since WIP

ekcoh commented:
Excellent with a GIF animation by the way quickly providing a good comprehension of what I am reviewing.

lyndon-unity commented:
> Excellent with a GIF animation by the way quickly providing a good comprehension of what I am reviewing.

Agreed although the default scheme, auto switch and default map doesn't seem present for the input action asset type if I'm interpretting the gif correctly. I'd expect them to be present for project wide input actions and non null input action assets.

jfreire-unity commented:
> > Excellent with a GIF animation by the way quickly providing a good comprehension of what I am reviewing.
> 
> Agreed although the default scheme, auto switch and default map doesn't seem present for the input action asset type if I'm interpretting the gif correctly. I'd expect them to be present for project wide input actions and non null input action assets.

That only happens if the actions asset has control schemes created. The one I selected in the GIF didn't have, but for instance, this one has:
<img width="588" alt="image" src="https://github.com/Unity-Technologies/InputSystem/assets/123556071/a8e9bd99-1f34-4d4a-bd69-90126ca50b7c">

This was the behavior that was already implemented by the way. I didn't change anything with regard to this area.

jfreire-unity committed:
Redo UI logic for InputActionAsset custom drawer

Avoids the introduction of a new class InputActionAssetProperty just for the sake of showing a custom property drawer.

jfreire-unity commented:
@ekcoh thanks for the idea of checking if the asset selected is a project-wide action or not. For some reason the first time I tried it did not work. I must have tried it in a rush and failed to get a simpler solution.

jfreire-unity commented:
@ekcoh thanks for the idea of checking if the asset selected is a project-wide action or not. It works! For some reason the first time I tried it did not work. I must have tried it in a rush and failed to get a simpler solution.

jfreire-unity committed:
Make custom property drawer available from 2022.3

jfreire-unity committed:
Make it non-public to avoid documentation errors

jfreire-unity committed:
Merge branch 'develop' into isx-1321-update-components-with-project-wide-actions

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/InputActionAssetDrawer.cs:
```diff
@@ -0,0 +1,71 @@
+            EditorGUI.BeginProperty(position, label, property);
+
+            var isAssetProjectWideActions = IsAssetProjectWideActions(property);
+            var selectedAssetOptionIndex = isAssetProjectWideActions ? 0 : 1;
```
Would be tidier if this was an enum rather than 0/1 int value

You can use EditorGUILayout.EnumPopup below to assist this

jfreire-unity replied:
Added those changes now

lyndon-unity reviewed (commented):
This looks like a simpler approach that is quite robust and would work in more places. Overall I'd approve but I think a change to use enums rather than int's for section would be easier to read.

jfreire-unity committed:
Changed asset options to be an enum instead of int

jfreire-unity committed:
Merge branch 'isx-1321-update-components-with-project-wide-actions' of github.com:Unity-Technologies/InputSystem into isx-1321-update-components-with-project-wide-actions

lyndon-unity reviewed (approved):
The enum changes make that easier to read - thank you

jfreire-unity committed:
Merge branch 'develop' into isx-1321-update-components-with-project-wide-actions

jfreire-unity committed:
Fix CI by pulling 2023.1/staging image

This is due to fixes introduced by https://github.cds.internal.unity3d.com/unity/unity/pull/35384 that at the moment are only available on the staging branch.

jfreire-unity committed:
Merge branch 'isx-1321-update-components-with-project-wide-actions' of github.com:Unity-Technologies/InputSystem into isx-1321-update-components-with-project-wide-actions

lyndon-unity commented:
Can you indent the asset when the "Actions Asset" is selected?

Both in the player input component and my own component its not obvious its related to the drop down when the next field is at the same indenting depth

![image](https://github.com/Unity-Technologies/InputSystem/assets/33493311/edb10ab2-de87-4b6e-9ba3-fb33295da8af)

In the player input component its a bit confusing that both fields are called 'Actions'. I think the name with the drop down should perhaps be taken from the field name and then the actual asset selector just called 'Asset'

(in the above example my field in the 'Digital and analog script' is called "Input Action Asset" so I'd expect to see that even when the selection is on "Project-Wide Actions")

lyndon-unity commented:
On a related note the InputActionReference I have is not listing the project wide input actions when it pops up its selector. We need to find a way to inject it here too, but that could be a separate PR

![image](https://github.com/Unity-Technologies/InputSystem/assets/33493311/c1f7cd51-b670-4976-8341-77a889bfb4f3)

Note changes in this area might clash with this PR
https://github.com/Unity-Technologies/InputSystem/pull/1726

jfreire-unity commented:
> Can you indent the asset when the "Actions Asset" is selected?
> 
> Both in the player input component and my own component its not obvious its related to the drop down when the next field is at the same indenting depth
> 
> ![image](https://user-images.githubusercontent.com/33493311/267988389-edb10ab2-de87-4b6e-9ba3-fb33295da8af.png)
> 
> In the player input component its a bit confusing that both fields are called 'Actions'. I think the name with the drop down should perhaps be taken from the field name and then the actual asset selector just called 'Asset'
> 
> (in the above example my field in the 'Digital and analog script' is called "Input Action Asset" so I'd expect to see that even when the selection is on "Project-Wide Actions")

Ah yeah, good catch. I had a different name before but that was lost during refactoring. Will add "Actions Asset" label to that field.

jfreire-unity committed:
Indent and rename actions asset label

lyndon-unity committed:
Fixed the InputSystemUIInputModuleEditor for null asset selection and missing actions within the asset

lyndon-unity committed:
Merge branch 'isx-1321-update-components-with-project-wide-actions' of github.com:Unity-Technologies/InputSystem into isx-1321-update-components-with-project-wide-actions

lyndon-unity committed:
Fixed formatting

lyndon-unity committed:
Added TrackedDevicePosition and TrackedDeviceOrientation into the project wide input actions defaults for the standard UI bindings

lyndon-unity committed:
Fixed a test after new actions added to project-wide input actions

jfreire-unity committed:
Add open button to project-wide actions

jfreire-unity commented:
Adding QA to see if there are any issues we didn't catch in our testing cc @stefanunity or @Pauliusd01

Pauliusd01 commented:
Would changing the naming of one of the dropdown options from "Actions Asset" to "Custom Actions Asset" be decent idea? I feel like we hear the words Project Wide Actions a lot and we know what it means specifically so we can tell the 2 options apart easily, but to the user they both might sound too similar? Just a thought I had 😅

Pauliusd01 commented:
https://github.com/Unity-Technologies/InputSystem/assets/54306142/8fe75706-37df-47f5-9465-3d774c5895a6

Got 2 questions/potential issues:

- If simply switching between Project Wide Actions and Actions Assets should the already set Actions asset be saved in the Actions Asset dropdown? (To avoid users losing their progress, however small that may be when using the dropdown)
- In the video above I switch between the dropdown and then undo, my character no longer works. Does that mean there's an issue with the undos and the dropdown?

Pauliusd01 commented:
> Unity_2023-09-15_14-12-58.mp4
> 
> Got 2 questions/potential issues:
> 
>     * If simply switching between Project Wide Actions and Actions Assets should the already set Actions asset be saved in the Actions Asset dropdown? (To avoid users losing their progress, however small that may be when using the dropdown)
> 
>     * In the video above I switch between the dropdown and then undo, my character no longer works. Does that mean there's an issue with the undos and the dropdown?

Disregard the undo issue I guess, I reproduced it without your changes as well by just switching actions assets on Develop

jfreire-unity commented:
> > Unity_2023-09-15_14-12-58.mp4
> > Got 2 questions/potential issues:
> > ```
> > * If simply switching between Project Wide Actions and Actions Assets should the already set Actions asset be saved in the Actions Asset dropdown? (To avoid users losing their progress, however small that may be when using the dropdown)
> > 
> > * In the video above I switch between the dropdown and then undo, my character no longer works. Does that mean there's an issue with the undos and the dropdown?
> > ```
> 
> Disregard the undo issue I guess, I reproduced it without your changes as well by just switching actions assets on Develop

Yeah, I was just checking it. Maybe we can open a bug for it @Pauliusd01 ?

Pauliusd01 reviewed (approved):
Looks good to me, I only have the 2 not super consequential nitpicks I listed previously. Here's what I checked:

1. Checking existing, previously set up objects with play input and made sure that they still work the same.
2. Set up new controls with player input and Project Wide Actions, also made sure my modifications to the Project Wide Actions are still picked up by my player input object I set up. Tested it with multiple control schemes set.
3. Checked undoing when switching between Project Wide Actions and custom ones (One of the nitpicks I have mentioned)

jfreire-unity commented:
Merging with un-related failures for 2023.1. Applying what #1750 did, doesn't solve the issue so we will need to wait for the `trunk` changes to be propagated.
